CIPs/0007.md

```diff
@@ -0,0 +1,108 @@
+- Status: WIP
+
+## Overview
+
+This proposal contains a new DowntimeSlasher contract.
+
+The actual DowntimeSlasher tries to verify that the validator signed at least a block in all the window that is required to be validated which it leads to running out of gas for big windows.
+
+The main goal of this new contract is to generate a tool that would allow any user to slash a validator for every possible block windows without gas limitations. 
+
+For the actual window of 1/2 epoch, the DowntimeSlasher won't finish with a "run out gas" exception. Without this new tool for slashing, there is no a possible way to penalize a validator for downtime.
+
+## Goals
+
+Make a new contract that:
+- Allows the user to slash a validator if it was down for a specific consecutive block window
+- The implementation works for every possible block window (required in the initialization)
+
+## Proposed Solution
+
+Create a new contract funtion that divides the downtime verification in valid slots.
+
+The user that wants to slash a particular validator will have to call a function from the contract `S_n` times (`S_n` could be asked to the contract itself) to verify each slot.
+
+Then the user must generate another contract call to verify that the validator to be slashed was down for every consecutive slot.
+
+### Process
+
+1) The user calls the contract to know how many slots are required (`S_n`) and how many blocks must have every slot (`B_n`)
+
+2) The user requires to validate `S_n` slots
+    - The user calls the contract for the slot validation `[StartBlockNumber + (i * B_n), StartBlockNumber + ((1 + i) * B_n) )` with `0 <= i < S_n`
+
+3) The user calls the slash method with the `StartBlockNumber` (internally will verify that every slot validation was performed and succeded)
+
+### Variation of the proposed solution
+
+- Every user generates the validation slots for itself. The first user to gather all the validations and trigger the slashing, collects the reward
+
+- The slots validated works for every user in the community. When a slashing is performed, every user that generated a validated slot that is used to demostrate the downtime, receives a portion of the reward 
+
+## Alternative Solutions
+
+### Validator's Uptime in the epoch
+
+The Celo blockchain for reward purposes, calculates an uptime for each validator. The validator’s uptime for the epoch is the proportion of blocks in the epoch for which it is ‘up’: U = counter / (epoch_size - 11).
+
+One possible approach, it's to slash the validators if they are not covering a certain umbral.
+
+#### Pros
+
+The uptime calculation it's part of the Celo protocol. This not only allow us to save gas from calculating a possible slashing, but also could let the same blockchain to perform the slashes. This would make a safer net, instead of requiring a so rigurous control of the users.
+
+In the actual scenario, if no user is controling, no one get slashed.
+
+#### Cons
```

Comment:
Seems like one way to address most (if not all the cons) is that every validator will check for uptime at *every* block, and not just at the end of the epoch.